Provide even more context to the lint context

bernt-matthias opened the pull request:
Follow up on https://github.com/galaxyproject/galaxy/pull/12978

- main improvement is the addition of the source file (the tool file or a macro file). Note that this is the full path (so in a planemo run this will be `/tmp/...`)
- Also added a mechanism to extent lint messages. 
- In order to make this usable from planemo or language server I added `get_lint_context_for_tool_source` and changed `lint_tool_source_with` and `lint_xml_with` to return the lint context. 

Also added a first test for linting yaml and cwl tools (and fixed minor bugs)

## How to test the changes?
(Select all options that apply)
- [x] I've included appropriate [automated tests](https://docs.galaxyproject.org/en/latest/dev/writing_tests.html).
- [ ] This is a refactoring of components with existing test coverage.
- [ ] Instructions for manual testing are as follows:
  1. [add testing steps and prerequisites here if you didn't write automated tests covering all your changes]

## License
- [x] I agree to license these contributions under [Galaxy's current license](https://github.com/galaxyproject/galaxy/blob/dev/LICENSE.txt).
- [x] I agree to allow the Galaxy committers to license these and all my past contributions to the core galaxy codebase under the [MIT license](https://opensource.org/licenses/MIT). If this condition is an issue, uncheck and just let us know why with an e-mail to galaxy-committers@lists.galaxyproject.org.

bernt-matthias committed:
small restructuring of linter unit tests

bernt-matthias committed:
lint for invalid regex in stdio regex

bernt-matthias committed:
two minor improvements to cwl and yaml tools

in preparation to get linter tests running for such tools

cwl:

- remove a print from SchemaLoader

yaml:

- define YamlToolSource.parse_tool_type otherwise yaml tools are parsed
  like xml tools
- parse_version: return a string, otherwise 1.0 is treated as float
  which makes the linter (and probably more) fail

bernt-matthias committed:
fix node.base to macro file

while parsing macros

bernt-matthias committed:
xsd: make name attribute required

bernt-matthias committed:
add file name to linter message

- add fname to LintMessage
- LintContext make lint messages persistent over multiple linter calls
  before the messages were reset before the call of the next linter
  which is "impractical" if we want to use the lintmessages eg in
  planemo
- add test for linting a tool + (internal and external) macros
  to ensure that the fname and line number is correct
- add a simple linter test for yaml and json tools

bernt-matthias committed:
revert str function for lint messages

such that the original bahavior is unchanged
and add pretty print for use in external apps like planemo

bernt-matthias committed:
fix deprecation warning

if general_node is not None that this syntax is deprecated
(would return True if the node has no children)

mvdbeek commented on lib/galaxy/tool_util/parser/yaml.py:
```diff
@@ -36,11 +36,14 @@ def __init__(self, root_dict, source_path=None):
         return self._source_path
 
+    def parse_tool_type(self):
+        return 'yml'
```
```suggestion
        return self.root_dict.get("tool_type")
```

mvdbeek replied:
I think that should fix the failing tests, but I'll push the milestone for now.

bernt-matthias committed:
Fix yaml tool type

Co-authored-by: Marius van den Beek <m.vandenbeek@gmail.com>

mvdbeek commented on lib/galaxy/tool_util/linters/stdio.py:
```diff
@@ -1,29 +1,27 @@
-        tool_line = 0
-        tool_path = None
+    # determine node to report for general problems with stdio
+    tool_node = tool_xml.find("./tool")
```
We're allowing `tool_xml` to be `None` here, so I guess you want to guard this with a `if tool_xml is not None:` ?

bernt-matthias replied:
I would suggest to move the linter tests for yaml and cwl to a separate PR. 

The problem here is that the main lint function should distinguish which linters to call based on the tool language and not the tool type. Thus, all the xml linters are called for the yaml tools.

Also with the suggested change the tool_type is still `None` (because the yaml file does not include it).

mvdbeek commented on test/unit/tool_util/test_tool_linters.py:
```diff
@@ -739,7 +764,7 @@
             and "Select parameter of a conditional [select] options have to be defined by 'option' children elements." in x.error_messages
```
Just a heads up, I've rebased the branch and I'm rewriting these tests as standard tests. The pytest parameterization makes it very hard to tell what has failed.

bernt-matthias replied:
@mvdbeek thanks. Also just rebased and just wanted to push. Should I wait?

mvdbeek replied:
Yes, I think that would be better

bernt-matthias replied:
Cool. Just ping me. 

The new linting for assertions will need to get the line and xpath added to the `lint_ctx.error` calls. Also in my rebase I got the repeat linter test duplicated.

mvdbeek replied:
I think it's better to pass the parameter to the error call, and have the lint_ctx accept a function that gets you the lines. That way you can adapt this to any tool language by providing a different function, and you don't need to call `node_props` everywhere

mvdbeek replied:
Hmm, so what i've got so far is in https://github.com/galaxyproject/galaxy/compare/dev...mvdbeek:topic/lint-context2?expand=1. Does that look good to you and would you have a chance to finish rewriting the tests ?

bernt-matthias replied:
> Does that look good to you and would you have a chance to finish rewriting the tests ?

Sure. 

> I think it's better to pass the parameter to the error call, and have the lint_ctx accept a function that gets you the lines. 
That way you can adapt this to any tool language by providing a different function, and you don't need to call node_props everywhere

Sounds good. But I have two questions:

- xml tools this seems easy for for source line, but the xpath requires the ETree object .. in external applications like planemo / GLS this will be a different object (built from the same source) .. maybe the xpath lookup won't work then. But I guess xpath is not so important anyway
- in case of cwl tools a parameter would be a dict, or? I have no clue how one can deduce something like the source line from the a dict, i.e. part of the parsed json?

mvdbeek replied:
Hmm, this was just an idea, maybe it's fine the way it is. I don't know that we really need linting for CWL, the tool loader already does a lot of validation via the schema. If we're investing in that we should first discuss that with the CWL community.

mvdbeek committed:
Bind tool_xml to node_props, so we don't have to pass it around

mvdbeek committed:
Rewrite parameterized pytests to regular test

The .. and  ... and ... and within a lambda gives you
no chance to debug the test without a print statement or
a debugger.

Co-authored-by: Marius van den Beek <m.vandenbeek@gmail.com>

bernt-matthias committed:
add function to return lint context

this adds functionality to the linter functions to return the lint
context. without breaking previous behavior.

Planemo seems to use

- lint_tool_source (which returned the fail state) and
- lint_tool_source_with (which returned nothing)

For the later a parameter has been added to make the linter silent
and the lint context is now returned.

For the former a variant has been added which returns the lint context.

bernt-matthias committed:
adapt tool loader test

when loading macros from imported files
the base property of the node is
overwritten.

this adds a xml:base property to the node
which is printed in the test

mvdbeek commented on lib/galaxy/tool_util/lint.py:
```diff
@@ -89,59 +124,67 @@ def __init__(self, level, skip_types=None, object_name=None):
+        if not silent:
+            # this is a relict from the past where the lint context
+            # was reset when called with a new lint_func, as workaround
+            # we safe the message list, apply the lint_func (which then
```
```suggestion
            # we save the message list, apply the lint_func (which then
```

davelopez commented on lib/galaxy/tool_util/lint.py:
```diff
@@ -12,15 +12,36 @@
     return not lint_context.failed(fail_level)
 
 
+def lint_context_for_tool_source(tool_source, extra_modules=None, skip_types=None, name=None):
```
I would probably change the signature a bit like this to be more explicit about what we are returning. But for that, we need to move the declaration of LintContext and LintMessage before this or use forward declaration like `-> "LintContext"`

```suggestion
def get_lint_context_for_tool_source(tool_source, extra_modules=None, skip_types=None, name=None) -> LintContext:
```

davelopez commented on lib/galaxy/tool_util/lint.py:
```diff
@@ -89,59 +124,67 @@ def __init__(self, level, skip_types=None, object_name=None):
+        self.message_list = []
+
+    @property
+    def found_errors(self):
```
All these properties are easy to annotate like:
```suggestion
    def found_errors(self) -> bool:
```
Not a big deal but it would be nice :)

davelopez commented on lib/galaxy/tool_util/lint.py:
```diff
@@ -89,59 +124,67 @@ def __init__(self, level, skip_types=None, object_name=None):
         self.object_name = object_name
-        self.found_errors = False
-        self.found_warns = False
+        self.message_list = []
```
```suggestion
        self.message_list: List[LintMessage] = []
```
For this, we will need to import `from typing import List` but again it is nicer to be explicit about the type here

bernt-matthias replied:
I added type annotations at most places. Maybe you can recheck?

davelopez replied:
Thank you! That's a lot of type hints :laughing: 
Check `mypy` errors, usually, they give a nice hint to fix them.
I would use just `None` instead of `NoReturn` for methods. 
As per the documentation:
> `NoReturn` is a special type to annotate functions that never return normally. For example, a function that unconditionally raises an exception.

davelopez commented on lib/galaxy/tool_util/lint.py:
```diff
@@ -89,59 +124,67 @@ def __init__(self, level, skip_types=None, object_name=None):
+                for message in self.valid_messages:
+                    plf = print_linter_info(plf)
+                    print(f"{message}")
+            self.message_list = tmp_message_list + self.message_list
```
I think `tmp_message_list` can be unbound here if we don't declare it as an empty list before line 144?

bernt-matthias replied:
Its both times `if not silent` .. so it should be fine. But if you prefer I can add the extra line.

davelopez replied:
You are right, I guess there's no need for the extra line

davelopez commented on lib/galaxy/tool_util/lint.py:
```diff
@@ -158,22 +201,23 @@ def warn_messages(self):
     def error_messages(self):
```
I think it would be more useful to return here the actual `List[LintMessage]` with all the metadata instead of just the string message. The consumer of this property can then make use of the metadata if required.

bernt-matthias replied:
Done.

davelopez reviewed (commented):
> could you have a look if this is useful 

Yeah! this is what I had in mind, I can't wait to try and integrate it in GLS :)

I just added some comments mostly about typing, but they are just minor improvements.

bernt-matthias committed:
add type annotations

bernt-matthias committed:
refactoring of LintMessage

subclassed XMLLintMessage

bernt-matthias committed:
eliminate node_props

bernt-matthias committed:
more linting refactoring

Special LintMessage for XML:

- one adding line and file name and
- one adding XPath

added error and report level class

also added doc string

bernt-matthias committed:
Fix typo

Co-authored-by: Marius van den Beek <m.vandenbeek@gmail.com>
